### PR TITLE
Three

### LINKED ISSUE #45: Rethink automatic starting of simulations?
Reading the [Force-Directed Graph](http://bl.ocks.org/mbostock/f584aa36df54c451c94a9d0798caed35) example, I find it hard to understand, _when_ the simulation starts.

According to the the API, "The simulator starts automatically". So that would be here:

```
var simulation = d3.forceSimulation()
    .force("link", d3.forceLink().id(function(d) { return d.id; }))
    .force("charge", d3.forceManyBody())
    .force("center", d3.forceCenter());
```

Hmmm, without nodes, there wont be much to simulate. Continue reading the code, I assume, it might happen here:

```
simulation
  .nodes(graph.nodes)
  .on("tick", ticked);
```

Now it has nodes and can work. But it still has no links. Hmmm. So does it run considering the other forces? Well, Javascript is singlethreaded, so the time callback cannot execute anyway. So presumably it the timer fires, but the callback will be run after this function, or more exactly, after the XmlHttpRequest callback is handled. Lots of pondering. Continue reading

```
simulation.force("link")
  .links(graph.links);
```

I assume that now the simulation has all arguments and can run considering the link forces. Which in fact is correct. Anyway, I find the API a bit confusing in this point (otherwise the new force API is absolutely ingenious!).

Debugging reveals, that force.initialize() is called 3 times, computing 

```
nodeById = map$1(nodes, id),
```

twice, unnecessarily. So maybe it makes sense to have an API method alike

```
simulation.start(nodes, links)
```

or more abstract (or precise)

```
simulation.start(all parameters required to run the configured simulation)
```

? For me that would have been clearer. Most functions would fall into category **configure** while _start_, _restart_, _stop_ would be category **run**.

thx so much for d3
urbanhop

### LINKED ISSUE #81: d3.forceCenter should have an alpha parameter.
My use case is that I'm building a UI where people can add a node by double-clicking. The node appears where the mouse is.
I tried using `d3.forceCenter`, but when the node appears, it just recenters everything instantly which I find a bit aggressive. I want to provide a smoother experience (the whole graph gradually recenters)

Filing this issue for various reasons:
* knowing whether this is the wanted behavior for default d3
* whether the d3 project is interested in a PR on this topic (if so let's discuss what the new behavior would be)
* if it's not a PR for the project, where should people gather if they want to create a library of forces for d3-force? Maybe just npm with the `d3-force` keyword?

### PR DESCRIPTION
**Pull request by Fil**
- [x] only start the simulation when someone is listening (fixes #45)
- [x] make forceCenter less aggressive  (fixes #81)
- [x] tests (#17) 
- [x] add forceRadial x, y and angle accessors (#167)

Because of the breaking changes in <del>#81 (which impacts #17),</del> #167 (and partly #45) this branch is "maximalist", and should be published as `d3-force@3`.

I would prefer to do this "maximalist" approach now to integrate into d3v6, but the alternative option is to have a minimalist d3-force@2 (#172) and wait for d3@7 to include these changes.

### CODE REVIEW

File: src/radial.js [hunk 1]
```diff
@@ -38,19 +72,23 @@ export default function(radius, x, y) {
...
-    return arguments.length ? (x = +_, force) : x;
+    return arguments.length ? (x = value(_) || constant(0), initialize(), force) : x;
   };
 
   force.y = function(_) {
-    return arguments.length ? (y = +_, force) : y;
+    return arguments.length ? (y = value(_) || constant(0), initialize(), force) : y;
+  };
+
+force.angle = function(_) {
```
Comment (mbostock) [inline src/radial.js:90]:
```suggestion
  force.angle = function(_) {
```

**Comment by mbostock** [comment]
If we’re willing to break backwards-compatibility then I think having explicit start is better than implicit start on the first event listener.

**Comment by Fil** [comment]
I've just pushed that change (with *simulation*.restart aliased as *simulation*.start).

The code is admittedly cleaner, but it feels a bit awkward. (Maybe I just need a moment to get used to it.)

**Comment by Fil** [comment]
Concerning forceCenter() there is probably a case to be made for when you want an immediate centering (in particular to recenter an initial phyllotaxis arrangement).

Should we remove the dependency on alpha and just leave an optional strength() function? Then the default strength(1) would be immediate, and strength(0.05) or so would be the go-to choice for gentle interactive recentering.

**Comment by Fil** [comment]
TL;DR: Drop #45 ; merge #167, #81 and #17.

After more consideration and tests:

-  I'm not terribly happy with #45; automatic start is nicer to use than having to explicitly call *simulation*.start, and easy to override with *simulation*.stop(). The logic of "waiting for listeners" (dubbed solution 1) was not right either (even if that's not recommended, you can imagine a situation in which the forces themselves are producing some side effects, the simulation being used like a timer…). The issue it wants to address can be solved by an example (stop, load then restart). So I'd vote to walk back this part.

- forceRadial x, y and angle accessors (#167) are great; I don't think that it will impact many examples (or any at all) but the change is technically breaking (the getters now returning a function rather than a number). We might want to add a strength setter too.

- forceCenter.strength solves #81 and defaults to the previous behavior, meaning it can be included with no breaking change, now or later.

- the tests are now immune to the change in forceCenter, so they can be included with no breaking change

**Comment by Fil** [comment]
closing in favor of #174
